src/adb.js

```diff
@@ -487,6 +525,134 @@ class Adb {
         );
     });
   }
+
+  // AW : return the file size of a complete folder
+ getFileSize(file) {
+  var _this = this;
+  return new Promise(function(resolve, reject) {
+  _this.getAdbState().then((stdout)=>{
+       if (stdout == 1) { // Recovery
+            _this.shell('du -shk '+file).then( (stdout)=>{
+                  stdout = parseFloat(stdout);
+                  _this.log("FileSize is "+stdout+ " Ko");
+                  resolve(stdout);
+            }).catch (reject);
+           }
+        else if (stdout == 2) { // device (type : used, size)
+            _this.shell('df -hBK '+file+' --output=used|tail -n1').then( (stdout)=>{
+                  stdout = parseFloat(stdout);
+                  _this.log("FileSize is "+stdout+ " Ko");
+                  resolve(stdout);
+            }).catch (reject);
+           }
+   });
+  }).catch (e=>{_this.log(e); reject;});
+ }
+
+  // AW : Return the status of the device (0 : bootloader, 1 : recovery, 2 : device ...)
+ getAdbState() {
+  var _this = this;
+  return new Promise(function(resolve, reject) {
+  _this.execCommand(["get-state"]).then( (stdout)=>{
+     _this.log("Device state is "+stdout);
+      switch (stdout) {
+             case "device"     : resolve(2);
+             case "bootloader" : resolve(0);
+             case "recovery"   : resolve(1);
+             default : resolve(3); // Unknown
+         }
+    }).catch (e=>{
+          // Try with recovery command
+          _this.log("Unknown error :"+e);     
+          reject
+         });
+  });
+ }
+
+// AW : Backup file "srcfile" from the phone to "destfile" localy
+// TO-DO : This function should be moved to devices.js
+ backup(srcfile,destfile, adbpath, progress) {
+    var _this = this;
+    _this.log("Backup Function Entry, will backup "+srcfile+" to "+destfile+" adbpath:"+adbpath);
+    return new Promise(function(resolve, reject) {
+        _this.fileSize = 9999999999999; // Init to file size to a very high value while waiting for the real size.
+        // Get file size
+        _this.getFileSize(srcfile).then(stdout=>{
+           _this.log("Returned FileSize is "+stdout+ " Ko"); 
+           _this.fileSize = stdout;
+           // Creating pipe
+           _this.shell("mkfifo /backup.pipe").then((stdout) => {
+              _this.log("Pipe created !");
+              var lastSize = 0;
+              var progressInterval = setInterval(() => {
+                 _this.execCommandCustom(["stat", "-t", destfile]).then(stat => {
```

Comment:
Use [the fs module](https://nodejs.org/api/fs.html#fs_fs_stat_path_options_callback) for this instead.

Reply:
done